contracts/src/PBHVerifier.sol

```diff
@@ -1,32 +1,154 @@
+    //
+    // - Initialisation and ownership management are not protected behind `onlyProxy` intentionally.
+    //   This ensures that the contract can safely be disposed of after it is no longer used.
+    // - Carefully consider what data recovery options are presented as new functionality is added.
+    //   Care must be taken to ensure that a migration plan can exist for cases where upgrades
+    //   cannot recover from an issue or vulnerability.
+
+    ///////////////////////////////////////////////////////////////////////////////
+    ///                    !!!!! DATA: DO NOT REORDER !!!!!                     ///
+    ///////////////////////////////////////////////////////////////////////////////
+
+    // To ensure compatibility between upgrades, it is exceedingly important that no reordering of
+    // these variables takes place. If reordering happens, a storage clash will occur (effectively a
+    // memory safety error).
+
+    using ByteHasher for bytes;
+
     ///////////////////////////////////////////////////////////////////////////////
-    ///                             CONSTRUCTION                                ///
+    ///                                  ERRORS                                ///
+    //////////////////////////////////////////////////////////////////////////////
+
+    /// @notice Thrown when attempting to reuse a nullifier
+    error InvalidNullifier();
+
     ///////////////////////////////////////////////////////////////////////////////
+    ///                                  Events                                ///
+    //////////////////////////////////////////////////////////////////////////////
 
-    /// @notice Constructs a new instance of the PBH Verifier.
-    /// @dev This constructor is only called once, and can be called with the encoded call necessary
-    ///      to initialize the logic contract.
+    /// @notice Emitted once for each successful PBH verification.
     ///
-    /// @param _logic The initial implementation (delegate) of the contract that this acts as a proxy
-    ///        for.
-    /// @param _data If this is non-empty, it is used as the data for a `delegatecall` to `_logic`.
-    ///        This is usually an encoded function call, and allows for initialising the storage of
-    ///        the proxy in a way similar to a traditional solidity constructor.
-    constructor(address _logic, bytes memory _data) payable WorldIDProxy(_logic, _data) {
-        // !!!! DO NOT PUT PROGRAM LOGIC HERE !!!!
-        // It should go in the `initialize` function of the delegate instead.
+    /// @param sender The sender of this particular transaction or UserOp.
+    /// @param nonce Transaction/UserOp nonce.
+    /// @param callData Transaction/UserOp call data.
+    /// @param payload The zero-knowledge proof that demonstrates the claimer is registered with World ID.
+    event PBH(address indexed sender, uint256 indexed nonce, bytes callData, PBHPayload payload);
```

Comment:
Should we emit the `calldata`? It seems like it would be more efficient to emit the `signalHash` derived from the UserOp calldata?